// @genkit-ai/google-genai/lib/vertexai/interaction-converters.mjs
import { cleanSchema } from "./utils.mjs";
function ensureToolIds(messages) {
  const generatedIds = [];
  let nextIdCounter = 0;
  const newMessages = structuredClone(messages);
  for (const message of newMessages) {
    for (const part of message.content) {
      if (part.toolRequest && !part.toolRequest.ref) {
        const newId = `genkit-auto-id-${nextIdCounter++}`;
        part.toolRequest.ref = newId;
        generatedIds.push(newId);
      }
    }
  }
  for (const message of newMessages) {
    for (const part of message.content) {
      if (part.toolResponse && !part.toolResponse.ref) {
        const id = generatedIds.shift();
        if (id) {
          part.toolResponse.ref = id;
        } else {
          part.toolResponse.ref = `genkit-orphan-id-${nextIdCounter++}`;
        }
      }
    }
  }
  return newMessages;
}
function toInteractionTool(tool) {
  const func = {
    type: "function",
    name: tool.name,
    description: tool.description
  };
  if (tool.inputSchema) {
    func.parameters = cleanSchema(tool.inputSchema);
  }
  return func;
}
function toInteractionContent(part) {
  if (part.text !== void 0) {
    return { type: "text", text: part.text };
  }
  if (part.media) {
    return toInteractionMedia(part);
  }
  if (part.toolRequest) {
    return {
      type: "function_call",
      name: part.toolRequest.name,
      arguments: part.toolRequest.input,
      id: part.toolRequest.ref || ""
    };
  }
  if (part.toolResponse) {
    let output = part.toolResponse.output;
    if (typeof output !== "object" && typeof output !== "string" && output !== void 0) {
      output = { result: output };
    }
    return {
      type: "function_result",
      name: part.toolResponse.name,
      result: output,
      call_id: part.toolResponse.ref || ""
    };
  }
  throw new Error("Unsupported part type for Interaction input");
}
function toInteractionMedia(part) {
  if (!part.media) throw new Error("Media part missing media");
  const { url, contentType } = part.media;
  if (!contentType) throw new Error("Media part missing contentType");
  let data;
  let uri;
  if (url.startsWith("data:")) {
    data = url.substring(url.indexOf(",") + 1);
  } else {
    uri = url;
  }
  const out = { mime_type: contentType };
  if (data) out.data = data;
  if (uri) out.uri = uri;
  if (contentType.startsWith("image/")) {
    out.type = "image";
    return out;
  }
  if (contentType.startsWith("audio/")) {
    out.type = "audio";
    return out;
  }
  if (contentType.startsWith("video/")) {
    out.type = "video";
    return out;
  }
  if (contentType === "application/pdf") {
    out.type = "document";
    return out;
  }
  throw new Error(`Unsupported media type: ${contentType}`);
}
function toInteractionRole(role) {
  switch (role) {
    case "user":
      return "user";
    case "model":
      return "model";
    case "tool":
      return "user";
    case "system":
      throw new Error(
        `System role should be handled as system_instruction, not part of turns.`
      );
    default:
      return "user";
  }
}
function toInteractionTurn(message) {
  return {
    role: toInteractionRole(message.role),
    content: message.content.map(toInteractionContent)
  };
}
function fromInteractionContent(content) {
  switch (content.type) {
    case "text":
      return fromTextContent(content);
    case "image":
      return fromImageContent(content);
    case "audio":
    case "document":
      return fromMediaContent(content);
    case "video":
      return fromVideoContent(content);
    case "thought":
      return fromThoughtContent(content);
    case "function_call":
      return fromFunctionCallContent(content);
    case "function_result":
      return fromFunctionResultContent(content);
    default:
      throw new Error(
        `Unsupported content type: ${content.type}. Raw content: ${JSON.stringify(content)}`
      );
  }
}
function fromMediaContent(content) {
  let url = content.uri;
  if (content.data && content.mime_type) {
    url = `data:${content.mime_type};base64,${content.data}`;
  }
  return {
    media: {
      url: url || "",
      contentType: content.mime_type
    }
  };
}
function fromTextContent(content) {
  return {
    text: content.text || "",
    metadata: {
      annotations: content.annotations
    }
  };
}
function fromImageContent(content) {
  const part = fromMediaContent(content);
  if (content.resolution !== void 0) {
    part.metadata = { resolution: content.resolution };
  }
  return part;
}
function fromVideoContent(content) {
  const part = fromMediaContent(content);
  if (content.resolution !== void 0) {
    part.metadata = { resolution: content.resolution };
  }
  return part;
}
function fromThoughtContent(content) {
  let reasoning = "";
  if (content.summary) {
    reasoning = content.summary.map((c) => {
      if (c.type === "text") return c.text;
      return "[Image]";
    }).join("\n");
  }
  return {
    reasoning,
    metadata: {
      thoughtSignature: content.signature
    },
    custom: {
      thought: content
    }
  };
}
function fromFunctionCallContent(content) {
  return {
    toolRequest: {
      name: content.name,
      input: content.arguments,
      ref: content.id
    }
  };
}
function fromFunctionResultContent(content) {
  return {
    toolResponse: {
      name: content.name,
      output: content.result,
      ref: content.call_id
    }
  };
}
function fromInteractionSync(interaction) {
  if (interaction.status === "failed") {
    throw new Error("Interaction failed");
  }
  const response = {
    finishReason: "stop",
    message: {
      role: "model",
      content: []
    },
    custom: interaction,
    raw: interaction
  };
  if (interaction.status === "cancelled") {
    response.finishReason = "aborted";
    response.finishMessage = "Operation cancelled";
    response.message.content = [{ text: "Operation cancelled." }];
    return response;
  }
  const outputs = interaction.outputs;
  if (outputs?.length) {
    response.message.content = outputs.filter((o) => o && Object.keys(o).length > 0).map(fromInteractionContent);
    if (interaction.usage) {
      response.usage = {
        inputTokens: interaction.usage.total_input_tokens,
        outputTokens: interaction.usage.total_output_tokens,
        totalTokens: interaction.usage.total_tokens,
        cachedContentTokens: interaction.usage.total_cached_tokens,
        thoughtsTokens: interaction.usage.total_thought_tokens
      };
      if (interaction.usage.input_tokens_by_modality) {
        for (const modalityToken of interaction.usage.input_tokens_by_modality) {
          switch (modalityToken.modality) {
            case "text":
              response.usage.inputCharacters = modalityToken.tokens;
              break;
            case "image":
              response.usage.inputImages = modalityToken.tokens;
              break;
            case "audio":
              response.usage.inputAudioFiles = modalityToken.tokens;
              break;
          }
        }
      }
      if (interaction.usage.output_tokens_by_modality) {
        for (const modalityToken of interaction.usage.output_tokens_by_modality) {
          switch (modalityToken.modality) {
            case "text":
              response.usage.outputCharacters = modalityToken.tokens;
              break;
            case "image":
              response.usage.outputImages = modalityToken.tokens;
              break;
            case "audio":
              response.usage.outputAudioFiles = modalityToken.tokens;
              break;
          }
        }
      }
    }
  }
  return response;
}
function fromInteraction(interaction, clientOptions) {
  const op = { id: interaction.id };
  if (clientOptions) {
    op.metadata = { clientOptions };
  }
  if (interaction.status === "in_progress") {
    op.done = false;
  } else if (interaction.status === "cancelled") {
    op.done = true;
    op.output = {
      finishReason: "aborted",
      finishMessage: "Operation cancelled",
      message: {
        role: "model",
        content: [{ text: "Operation cancelled." }]
      }
    };
  } else if (interaction.status === "completed") {
    op.done = true;
    const outputs = interaction.outputs;
    if (outputs?.length) {
      const content = outputs.filter((o) => o && Object.keys(o).length > 0).map(fromInteractionContent);
      op.output = {
        finishReason: "stop",
        message: {
          role: "model",
          content
        },
        custom: interaction,
        raw: interaction
      };
      if (interaction.usage) {
        op.output.usage = {
          inputTokens: interaction.usage.total_input_tokens,
          outputTokens: interaction.usage.total_output_tokens,
          totalTokens: interaction.usage.total_tokens,
          cachedContentTokens: interaction.usage.total_cached_tokens,
          thoughtsTokens: interaction.usage.total_thought_tokens
        };
        if (interaction.usage.input_tokens_by_modality) {
          for (const modalityToken of interaction.usage.input_tokens_by_modality) {
            switch (modalityToken.modality) {
              case "text":
                op.output.usage.inputCharacters = modalityToken.tokens;
                break;
              case "image":
                op.output.usage.inputImages = modalityToken.tokens;
                break;
              case "audio":
                op.output.usage.inputAudioFiles = modalityToken.tokens;
                break;
            }
          }
        }
        if (interaction.usage.output_tokens_by_modality) {
          for (const modalityToken of interaction.usage.output_tokens_by_modality) {
            switch (modalityToken.modality) {
              case "text":
                op.output.usage.outputCharacters = modalityToken.tokens;
                break;
              case "image":
                op.output.usage.outputImages = modalityToken.tokens;
                break;
              case "audio":
                op.output.usage.outputAudioFiles = modalityToken.tokens;
                break;
            }
          }
        }
      }
    }
  }
  return op;
}
export {
  ensureToolIds,
  fromInteraction,
  fromInteractionContent,
  fromInteractionSync,
  toInteractionContent,
  toInteractionRole,
  toInteractionTool,
  toInteractionTurn
};
