// @genkit-ai/google-genai/lib/googleai/antigravity.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var antigravity_exports = {};
__export(antigravity_exports, {
  AntigravityConfigSchema: () => AntigravityConfigSchema,
  TEST_ONLY: () => TEST_ONLY,
  defineModel: () => defineModel,
  isAntigravityModelName: () => isAntigravityModelName,
  listActions: () => listActions,
  listKnownModels: () => listKnownModels,
  model: () => model
});
module.exports = __toCommonJS(antigravity_exports);
var import_genkit = require("genkit");
var import_model = require("genkit/model");
var import_plugin = require("genkit/plugin");
var import_utils = require("../common/utils.js");
var import_client = require("./client.js");
var import_interaction_converters = require("./interaction-converters.js");
var import_utils2 = require("./utils.js");
const AntigravityConfigSchema = import_genkit.z.object({
  apiKey: import_genkit.z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  previousInteractionId: import_genkit.z.string().describe("The ID of the previous interaction, if any.").optional(),
  store: import_genkit.z.boolean().describe(
    "Whether to store the response and request for later retrieval."
  ).optional(),
  environment: import_genkit.z.union([import_genkit.z.string(), import_genkit.z.record(import_genkit.z.unknown())]).describe("The environment configuration for the sandbox.").optional(),
  responseModalities: import_genkit.z.array(import_genkit.z.enum(["TEXT", "IMAGE"])).describe("The modalities to be used in response.").optional()
}).passthrough();
function commonRef(name, info, configSchema = AntigravityConfigSchema) {
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: true,
        media: true,
        tools: false,
        // function calling unsupported for antigravity
        toolChoice: false,
        systemRole: false,
        output: ["text"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("antigravity");
const KNOWN_MODELS = {
  "antigravity-preview-05-2026": commonRef("antigravity-preview-05-2026")
};
function isAntigravityModelName(value) {
  return !!value?.startsWith("antigravity-");
}
function model(version, config = {}) {
  const name = (0, import_utils2.checkModelName)(version);
  if ((0, import_utils.isKnownKey)(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    config,
    configSchema: AntigravityConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isAntigravityModelName((0, import_utils2.modelName)(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return (0, import_genkit.modelActionMetadata)({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  (0, import_utils2.checkApiKey)(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return (0, import_plugin.model)(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request, { abortSignal }) => {
      const clientOpt = (0, import_utils2.calculateRequestOptions)(
        { ...clientOptions, signal: abortSignal },
        request.config
      );
      const requestOptions = (0, import_utils2.removeClientOptionOverrides)(request.config || {});
      const {
        previousInteractionId,
        store,
        environment,
        responseModalities,
        ...rest
      } = requestOptions;
      const apiKey = (0, import_utils2.calculateApiKey)(
        pluginOptions?.apiKey,
        request.config?.apiKey
      );
      const messages = structuredClone(request.messages);
      for (const message of messages) {
        if (message.role === "system") {
          message.role = "user";
        }
      }
      let responseModalitiesConverted;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        agent: (0, import_utils2.extractVersion)(ref),
        input: (0, import_interaction_converters.toInteractionSteps)((0, import_interaction_converters.ensureToolIds)(messages)),
        previous_interaction_id: previousInteractionId,
        store,
        environment,
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await (0, import_client.createInteraction)(apiKey, req, clientOpt);
      return (0, import_interaction_converters.fromInteractionSync)(response);
    }
  );
}
const TEST_ONLY = { KNOWN_MODELS };
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AntigravityConfigSchema,
  TEST_ONLY,
  defineModel,
  isAntigravityModelName,
  listActions,
  listKnownModels,
  model
});
