// @genkit-ai/google-genai/lib/googleai/antigravity.mjs
import { modelActionMetadata, z } from "genkit";
import { modelRef } from "genkit/model";
import { model as pluginModel } from "genkit/plugin";
import { isKnownKey } from "../common/utils.mjs";
import { createInteraction } from "./client.mjs";
import {
  ensureToolIds,
  fromInteractionSync,
  toInteractionSteps
} from "./interaction-converters.mjs";
import {
  calculateApiKey,
  calculateRequestOptions,
  checkApiKey,
  checkModelName,
  extractVersion,
  modelName,
  removeClientOptionOverrides
} from "./utils.mjs";
const AntigravityConfigSchema = z.object({
  apiKey: z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  previousInteractionId: z.string().describe("The ID of the previous interaction, if any.").optional(),
  store: z.boolean().describe(
    "Whether to store the response and request for later retrieval."
  ).optional(),
  environment: z.union([z.string(), z.record(z.unknown())]).describe("The environment configuration for the sandbox.").optional(),
  responseModalities: z.array(z.enum(["TEXT", "IMAGE"])).describe("The modalities to be used in response.").optional()
}).passthrough();
function commonRef(name, info, configSchema = AntigravityConfigSchema) {
  return modelRef({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: true,
        media: true,
        tools: false,
        // function calling unsupported for antigravity
        toolChoice: false,
        systemRole: false,
        output: ["text"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("antigravity");
const KNOWN_MODELS = {
  "antigravity-preview-05-2026": commonRef("antigravity-preview-05-2026")
};
function isAntigravityModelName(value) {
  return !!value?.startsWith("antigravity-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (isKnownKey(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return modelRef({
    name: `googleai/${name}`,
    config,
    configSchema: AntigravityConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isAntigravityModelName(modelName(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return modelActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return pluginModel(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request, { abortSignal }) => {
      const clientOpt = calculateRequestOptions(
        { ...clientOptions, signal: abortSignal },
        request.config
      );
      const requestOptions = removeClientOptionOverrides(request.config || {});
      const {
        previousInteractionId,
        store,
        environment,
        responseModalities,
        ...rest
      } = requestOptions;
      const apiKey = calculateApiKey(
        pluginOptions?.apiKey,
        request.config?.apiKey
      );
      const messages = structuredClone(request.messages);
      for (const message of messages) {
        if (message.role === "system") {
          message.role = "user";
        }
      }
      let responseModalitiesConverted;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        agent: extractVersion(ref),
        input: toInteractionSteps(ensureToolIds(messages)),
        previous_interaction_id: previousInteractionId,
        store,
        environment,
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await createInteraction(apiKey, req, clientOpt);
      return fromInteractionSync(response);
    }
  );
}
const TEST_ONLY = { KNOWN_MODELS };
export {
  AntigravityConfigSchema,
  TEST_ONLY,
  defineModel,
  isAntigravityModelName,
  listActions,
  listKnownModels,
  model
};
