// @genkit-ai/google-genai/lib/googleai/deep-research.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var deep_research_exports = {};
__export(deep_research_exports, {
  DeepResearchConfigSchema: () => DeepResearchConfigSchema,
  TEST_ONLY: () => TEST_ONLY,
  defineModel: () => defineModel,
  isDeepResearchModelName: () => isDeepResearchModelName,
  listActions: () => listActions,
  listKnownModels: () => listKnownModels,
  model: () => model
});
module.exports = __toCommonJS(deep_research_exports);
var import_genkit = require("genkit");
var import_model = require("genkit/model");
var import_plugin = require("genkit/plugin");
var import_utils = require("../common/utils.js");
var import_client = require("./client.js");
var import_interaction_converters = require("./interaction-converters.js");
var import_utils2 = require("./utils.js");
const DeepResearchConfigSchema = import_genkit.z.object({
  apiKey: import_genkit.z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  thinkingSummaries: import_genkit.z.enum(["AUTO", "NONE"]).describe("Whether to include thought summaries in the response.").optional(),
  previousInteractionId: import_genkit.z.string().describe("The ID of the previous interaction, if any.").optional(),
  store: import_genkit.z.boolean().describe(
    "Whether to store the response and request for later retrieval."
  ).optional(),
  responseModalities: import_genkit.z.array(import_genkit.z.enum(["TEXT", "IMAGE", "AUDIO"])).describe("The modalities to be used in response.").optional(),
  visualization: import_genkit.z.enum(["AUTO", "OFF"]).describe("Whether to enable agent-generated charts and images.").optional(),
  collaborativePlanning: import_genkit.z.boolean().describe(
    "Whether to enable multi-turn plan review before research begins."
  ).optional(),
  // TODO(V2): Remove googleSearch, urlContext, and codeExecution wrappers
  // and rely entirely on `config.tools` being passed through.
  googleSearch: import_genkit.z.union([import_genkit.z.boolean(), import_genkit.z.object({}).passthrough()]).describe(
    "Search the public web. Enabled by default if no tools are provided."
  ).optional(),
  urlContext: import_genkit.z.union([import_genkit.z.boolean(), import_genkit.z.object({}).passthrough()]).describe(
    "Read and summarize web page content. Enabled by default if no tools are provided."
  ).optional(),
  codeExecution: import_genkit.z.union([import_genkit.z.boolean(), import_genkit.z.object({}).passthrough()]).describe(
    "Execute code to perform calculations and data analysis. Enabled by default if no tools are provided."
  ).optional(),
  fileSearch: import_genkit.z.object({
    fileSearchStoreNames: import_genkit.z.array(import_genkit.z.string())
  }).passthrough().optional(),
  mcpServers: import_genkit.z.array(
    import_genkit.z.object({
      name: import_genkit.z.string().optional(),
      url: import_genkit.z.string().optional(),
      headers: import_genkit.z.record(import_genkit.z.string()).optional(),
      allowedTools: import_genkit.z.array(import_genkit.z.string()).optional()
    }).passthrough()
  ).optional()
}).passthrough();
function commonRef(name, info, configSchema = DeepResearchConfigSchema) {
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: true,
        media: false,
        tools: false,
        toolChoice: false,
        systemRole: false,
        output: ["text"],
        longRunning: true
      }
    }
  });
}
const GENERIC_MODEL = commonRef("deep-research");
const ADVANCED_DEEP_RESEARCH_INFO = {
  supports: {
    multiturn: true,
    media: true,
    tools: true,
    toolChoice: false,
    systemRole: false,
    output: ["text", "media"],
    longRunning: true
  }
};
const KNOWN_MODELS = {
  "deep-research-pro-preview-12-2025": commonRef(
    "deep-research-pro-preview-12-2025"
  ),
  "deep-research-preview-04-2026": commonRef(
    "deep-research-preview-04-2026",
    ADVANCED_DEEP_RESEARCH_INFO
  ),
  "deep-research-max-preview-04-2026": commonRef(
    "deep-research-max-preview-04-2026",
    ADVANCED_DEEP_RESEARCH_INFO
  )
};
function isDeepResearchModelName(value) {
  return !!value?.startsWith("deep-research-");
}
function model(version, config = {}) {
  const name = (0, import_utils2.checkModelName)(version);
  if ((0, import_utils.isKnownKey)(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    config,
    configSchema: DeepResearchConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isDeepResearchModelName((0, import_utils2.modelName)(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return (0, import_genkit.modelActionMetadata)({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  (0, import_utils2.checkApiKey)(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return (0, import_plugin.backgroundModel)({
    name: ref.name,
    ...ref.info,
    configSchema: ref.configSchema,
    async start(request) {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        thinkingSummaries,
        visualization,
        collaborativePlanning,
        previousInteractionId,
        store,
        responseModalities,
        googleSearch,
        urlContext,
        codeExecution,
        fileSearch,
        mcpServers,
        ...rest
      } = request.config || {};
      const apiKey = (0, import_utils2.calculateApiKey)(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion
      };
      const messages = structuredClone(request.messages);
      for (const message of messages) {
        if (message.role === "system") {
          message.role = "user";
        }
      }
      const tools = [];
      if (request.tools?.length) {
        for (const tool of request.tools) {
          tools.push((0, import_interaction_converters.toInteractionTool)(tool));
        }
      }
      if (googleSearch) {
        tools.push({
          type: "google_search",
          ...googleSearch === true ? {} : googleSearch
        });
      }
      if (urlContext) {
        tools.push({
          type: "url_context",
          ...urlContext === true ? {} : urlContext
        });
      }
      if (codeExecution) {
        tools.push({
          type: "code_execution",
          ...codeExecution === true ? {} : codeExecution
        });
      }
      if (fileSearch) {
        const { fileSearchStoreNames, ...restFileSearch } = fileSearch;
        tools.push({
          type: "file_search",
          file_search_store_names: fileSearchStoreNames,
          ...restFileSearch
        });
      }
      if (mcpServers) {
        for (const mcpServer of mcpServers) {
          const { allowedTools, ...restMcp } = mcpServer;
          tools.push({
            type: "mcp_server",
            ...restMcp,
            ...allowedTools ? { allowed_tools: allowedTools } : {}
          });
        }
      }
      let responseFormat = void 0;
      if (request.output?.format === "json" || request.output?.contentType === "application/json") {
        responseFormat = {
          type: "text",
          mime_type: "application/json"
        };
        if (request.output.schema) {
          responseFormat.schema = (0, import_utils.cleanSchema)(request.output.schema);
        }
      }
      let responseModalitiesConverted;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        agent: (0, import_utils2.extractVersion)(ref),
        input: (0, import_interaction_converters.toInteractionSteps)((0, import_interaction_converters.ensureToolIds)(messages)),
        background: true,
        agent_config: {
          type: "deep-research",
          thinking_summaries: thinkingSummaries?.toLowerCase(),
          visualization: visualization?.toLowerCase(),
          collaborative_planning: collaborativePlanning
        },
        previous_interaction_id: previousInteractionId,
        store,
        tools: tools.length ? tools : void 0,
        response_format: responseFormat,
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await (0, import_client.createInteraction)(apiKey, req, newClientOptions);
      return (0, import_interaction_converters.fromInteraction)(response, newClientOptions);
    },
    async check(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || (0, import_utils2.calculateApiKey)(pluginOptions?.apiKey, void 0);
      const checkOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await (0, import_client.getInteraction)(apiKey, operation.id, checkOptions);
      return (0, import_interaction_converters.fromInteraction)(response, checkOptions);
    },
    async cancel(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || (0, import_utils2.calculateApiKey)(pluginOptions?.apiKey, void 0);
      const cancelOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await (0, import_client.cancelInteraction)(
        apiKey,
        operation.id,
        cancelOptions
      );
      return (0, import_interaction_converters.fromInteraction)(response, cancelOptions);
    }
  });
}
const TEST_ONLY = { KNOWN_MODELS };
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DeepResearchConfigSchema,
  TEST_ONLY,
  defineModel,
  isDeepResearchModelName,
  listActions,
  listKnownModels,
  model
});
