// @genkit-ai/google-genai/lib/googleai/deep-research.mjs
import { modelActionMetadata, z } from "genkit";
import {
  modelRef
} from "genkit/model";
import { backgroundModel as pluginBackgroundModel } from "genkit/plugin";
import { cleanSchema, isKnownKey } from "../common/utils.mjs";
import {
  cancelInteraction,
  createInteraction,
  getInteraction
} from "./client.mjs";
import {
  ensureToolIds,
  fromInteraction,
  toInteractionSteps,
  toInteractionTool
} from "./interaction-converters.mjs";
import {
  calculateApiKey,
  checkApiKey,
  checkModelName,
  extractVersion,
  modelName
} from "./utils.mjs";
const DeepResearchConfigSchema = z.object({
  apiKey: z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  thinkingSummaries: z.enum(["AUTO", "NONE"]).describe("Whether to include thought summaries in the response.").optional(),
  previousInteractionId: z.string().describe("The ID of the previous interaction, if any.").optional(),
  store: z.boolean().describe(
    "Whether to store the response and request for later retrieval."
  ).optional(),
  responseModalities: z.array(z.enum(["TEXT", "IMAGE", "AUDIO"])).describe("The modalities to be used in response.").optional(),
  visualization: z.enum(["AUTO", "OFF"]).describe("Whether to enable agent-generated charts and images.").optional(),
  collaborativePlanning: z.boolean().describe(
    "Whether to enable multi-turn plan review before research begins."
  ).optional(),
  // TODO(V2): Remove googleSearch, urlContext, and codeExecution wrappers
  // and rely entirely on `config.tools` being passed through.
  googleSearch: z.union([z.boolean(), z.object({}).passthrough()]).describe(
    "Search the public web. Enabled by default if no tools are provided."
  ).optional(),
  urlContext: z.union([z.boolean(), z.object({}).passthrough()]).describe(
    "Read and summarize web page content. Enabled by default if no tools are provided."
  ).optional(),
  codeExecution: z.union([z.boolean(), z.object({}).passthrough()]).describe(
    "Execute code to perform calculations and data analysis. Enabled by default if no tools are provided."
  ).optional(),
  fileSearch: z.object({
    fileSearchStoreNames: z.array(z.string())
  }).passthrough().optional(),
  mcpServers: z.array(
    z.object({
      name: z.string().optional(),
      url: z.string().optional(),
      headers: z.record(z.string()).optional(),
      allowedTools: z.array(z.string()).optional()
    }).passthrough()
  ).optional()
}).passthrough();
function commonRef(name, info, configSchema = DeepResearchConfigSchema) {
  return modelRef({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: true,
        media: false,
        tools: false,
        toolChoice: false,
        systemRole: false,
        output: ["text"],
        longRunning: true
      }
    }
  });
}
const GENERIC_MODEL = commonRef("deep-research");
const ADVANCED_DEEP_RESEARCH_INFO = {
  supports: {
    multiturn: true,
    media: true,
    tools: true,
    toolChoice: false,
    systemRole: false,
    output: ["text", "media"],
    longRunning: true
  }
};
const KNOWN_MODELS = {
  "deep-research-pro-preview-12-2025": commonRef(
    "deep-research-pro-preview-12-2025"
  ),
  "deep-research-preview-04-2026": commonRef(
    "deep-research-preview-04-2026",
    ADVANCED_DEEP_RESEARCH_INFO
  ),
  "deep-research-max-preview-04-2026": commonRef(
    "deep-research-max-preview-04-2026",
    ADVANCED_DEEP_RESEARCH_INFO
  )
};
function isDeepResearchModelName(value) {
  return !!value?.startsWith("deep-research-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (isKnownKey(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return modelRef({
    name: `googleai/${name}`,
    config,
    configSchema: DeepResearchConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isDeepResearchModelName(modelName(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return modelActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return pluginBackgroundModel({
    name: ref.name,
    ...ref.info,
    configSchema: ref.configSchema,
    async start(request) {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        thinkingSummaries,
        visualization,
        collaborativePlanning,
        previousInteractionId,
        store,
        responseModalities,
        googleSearch,
        urlContext,
        codeExecution,
        fileSearch,
        mcpServers,
        ...rest
      } = request.config || {};
      const apiKey = calculateApiKey(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion
      };
      const messages = structuredClone(request.messages);
      for (const message of messages) {
        if (message.role === "system") {
          message.role = "user";
        }
      }
      const tools = [];
      if (request.tools?.length) {
        for (const tool of request.tools) {
          tools.push(toInteractionTool(tool));
        }
      }
      if (googleSearch) {
        tools.push({
          type: "google_search",
          ...googleSearch === true ? {} : googleSearch
        });
      }
      if (urlContext) {
        tools.push({
          type: "url_context",
          ...urlContext === true ? {} : urlContext
        });
      }
      if (codeExecution) {
        tools.push({
          type: "code_execution",
          ...codeExecution === true ? {} : codeExecution
        });
      }
      if (fileSearch) {
        const { fileSearchStoreNames, ...restFileSearch } = fileSearch;
        tools.push({
          type: "file_search",
          file_search_store_names: fileSearchStoreNames,
          ...restFileSearch
        });
      }
      if (mcpServers) {
        for (const mcpServer of mcpServers) {
          const { allowedTools, ...restMcp } = mcpServer;
          tools.push({
            type: "mcp_server",
            ...restMcp,
            ...allowedTools ? { allowed_tools: allowedTools } : {}
          });
        }
      }
      let responseFormat = void 0;
      if (request.output?.format === "json" || request.output?.contentType === "application/json") {
        responseFormat = {
          type: "text",
          mime_type: "application/json"
        };
        if (request.output.schema) {
          responseFormat.schema = cleanSchema(request.output.schema);
        }
      }
      let responseModalitiesConverted;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        agent: extractVersion(ref),
        input: toInteractionSteps(ensureToolIds(messages)),
        background: true,
        agent_config: {
          type: "deep-research",
          thinking_summaries: thinkingSummaries?.toLowerCase(),
          visualization: visualization?.toLowerCase(),
          collaborative_planning: collaborativePlanning
        },
        previous_interaction_id: previousInteractionId,
        store,
        tools: tools.length ? tools : void 0,
        response_format: responseFormat,
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await createInteraction(apiKey, req, newClientOptions);
      return fromInteraction(response, newClientOptions);
    },
    async check(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || calculateApiKey(pluginOptions?.apiKey, void 0);
      const checkOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await getInteraction(apiKey, operation.id, checkOptions);
      return fromInteraction(response, checkOptions);
    },
    async cancel(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || calculateApiKey(pluginOptions?.apiKey, void 0);
      const cancelOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await cancelInteraction(
        apiKey,
        operation.id,
        cancelOptions
      );
      return fromInteraction(response, cancelOptions);
    }
  });
}
const TEST_ONLY = { KNOWN_MODELS };
export {
  DeepResearchConfigSchema,
  TEST_ONLY,
  defineModel,
  isDeepResearchModelName,
  listActions,
  listKnownModels,
  model
};
