// @genkit-ai/google-genai/lib/googleai/interaction-converters.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var interaction_converters_exports = {};
__export(interaction_converters_exports, {
  ensureToolIds: () => ensureToolIds,
  fromCodeExecutionCall: () => fromCodeExecutionCall,
  fromCodeExecutionResult: () => fromCodeExecutionResult,
  fromGoogleSearchCall: () => fromGoogleSearchCall,
  fromGoogleSearchResult: () => fromGoogleSearchResult,
  fromInteraction: () => fromInteraction,
  fromInteractionContent: () => fromInteractionContent,
  fromInteractionStep: () => fromInteractionStep,
  fromInteractionSync: () => fromInteractionSync,
  fromServerFunctionCall: () => fromServerFunctionCall,
  fromServerFunctionResult: () => fromServerFunctionResult,
  toInteractionContent: () => toInteractionContent,
  toInteractionRole: () => toInteractionRole,
  toInteractionSteps: () => toInteractionSteps,
  toInteractionTool: () => toInteractionTool
});
module.exports = __toCommonJS(interaction_converters_exports);
var import_genkit = require("genkit");
var import_utils = require("./utils.js");
function ensureToolIds(messages) {
  const generatedIds = [];
  let nextIdCounter = 0;
  const newMessages = structuredClone(messages);
  for (const message of newMessages) {
    for (const part of message.content) {
      if (part.toolRequest && !part.toolRequest.ref) {
        const newId = `genkit-auto-id-${nextIdCounter++}`;
        part.toolRequest.ref = newId;
        generatedIds.push(newId);
      }
    }
  }
  for (const message of newMessages) {
    for (const part of message.content) {
      if (part.toolResponse && !part.toolResponse.ref) {
        const id = generatedIds.shift();
        if (id) {
          part.toolResponse.ref = id;
        } else {
          part.toolResponse.ref = `genkit-orphan-id-${nextIdCounter++}`;
        }
      }
    }
  }
  return newMessages;
}
function toInteractionTool(tool) {
  const func = {
    type: "function",
    name: tool.name,
    description: tool.description
  };
  if (tool.inputSchema) {
    func.parameters = (0, import_utils.cleanSchema)(tool.inputSchema);
  }
  return func;
}
function toInteractionContent(part) {
  if (part.text !== void 0) {
    return { type: "text", text: part.text };
  }
  if (part.media) {
    return toInteractionMedia(part);
  }
  console.warn(
    `Unsupported part type for Interaction input: ${JSON.stringify(part)}`
  );
  return void 0;
}
function toInteractionMedia(part) {
  if (!part.media) throw new Error("Media part missing media");
  const { url, contentType } = part.media;
  if (!contentType) throw new Error("Media part missing contentType");
  let data;
  let uri;
  if (url.startsWith("data:")) {
    data = url.substring(url.indexOf(",") + 1);
  } else {
    uri = url;
  }
  const out = { mime_type: contentType };
  if (data) out.data = data;
  if (uri) out.uri = uri;
  if (contentType.startsWith("image/")) {
    out.type = "image";
    return out;
  }
  if (contentType.startsWith("audio/")) {
    out.type = "audio";
    return out;
  }
  if (contentType.startsWith("video/")) {
    out.type = "video";
    return out;
  }
  if (contentType === "application/pdf") {
    out.type = "document";
    return out;
  }
  throw new Error(`Unsupported media type: ${contentType}`);
}
function toInteractionRole(role) {
  switch (role) {
    case "user":
      return "user";
    case "model":
      return "model";
    case "tool":
      return "user";
    case "system":
      throw new Error(
        `System role should be handled as system_instruction, not part of turns.`
      );
    default:
      return "user";
  }
}
const GoogleSearchArgsSchema = import_genkit.z.object({ queries: import_genkit.z.array(import_genkit.z.string()) });
const RecordUnknownSchema = import_genkit.z.record(import_genkit.z.unknown());
const RecordUnknownOrStringSchema = import_genkit.z.union([RecordUnknownSchema, import_genkit.z.string()]);
const OptionalStringSchema = import_genkit.z.string().optional();
const GoogleSearchCallSchema = import_genkit.z.object({
  id: import_genkit.z.string(),
  arguments: GoogleSearchArgsSchema
});
const GoogleSearchResultSchema = import_genkit.z.object({
  callId: import_genkit.z.string(),
  result: RecordUnknownSchema
});
const ExecutableCodeSchema = import_genkit.z.object({
  code: import_genkit.z.string(),
  language: import_genkit.z.string().default("PYTHON")
});
const CodeExecutionResultSchema = import_genkit.z.object({
  output: import_genkit.z.string(),
  outcome: import_genkit.z.string().optional()
});
function toInteractionSteps(messages) {
  const steps = [];
  for (const message of messages) {
    const normalContent = [];
    for (const part of message.content) {
      if (part.toolRequest) {
        steps.push({
          type: "function_call",
          name: part.toolRequest.name,
          arguments: RecordUnknownSchema.optional().parse(
            part.toolRequest.input
          ),
          id: part.toolRequest.ref || ""
        });
      } else if (part.toolResponse) {
        let output = part.toolResponse.output;
        if (typeof output !== "object" && typeof output !== "string" && output !== void 0) {
          output = { result: output };
        }
        steps.push({
          type: "function_result",
          name: part.toolResponse.name,
          result: RecordUnknownOrStringSchema.optional().parse(output),
          call_id: part.toolResponse.ref || ""
        });
      } else if (part.custom?.googleSearchCall) {
        const gsCall = GoogleSearchCallSchema.parse(
          part.custom.googleSearchCall
        );
        steps.push({
          type: "google_search_call",
          id: gsCall.id,
          arguments: gsCall.arguments,
          signature: OptionalStringSchema.parse(
            part.metadata?.thoughtSignature
          )
        });
      } else if (part.custom?.googleSearchResult) {
        const gsResult = GoogleSearchResultSchema.parse(
          part.custom.googleSearchResult
        );
        steps.push({
          type: "google_search_result",
          call_id: gsResult.callId,
          result: gsResult.result,
          signature: OptionalStringSchema.parse(
            part.metadata?.thoughtSignature
          )
        });
      } else if (part.custom?.executableCode) {
        const execCode = ExecutableCodeSchema.parse(part.custom.executableCode);
        steps.push({
          type: "code_execution_call",
          id: import_genkit.z.string().parse(part.metadata?.callId),
          arguments: {
            code: execCode.code,
            language: execCode.language
          },
          signature: OptionalStringSchema.parse(
            part.metadata?.thoughtSignature
          )
        });
      } else if (part.custom?.codeExecutionResult) {
        const execResult = CodeExecutionResultSchema.parse(
          part.custom.codeExecutionResult
        );
        steps.push({
          type: "code_execution_result",
          call_id: import_genkit.z.string().parse(part.metadata?.callId),
          result: execResult.output,
          signature: OptionalStringSchema.parse(
            part.metadata?.thoughtSignature
          )
        });
      } else if (part.reasoning) {
        steps.push({
          type: "thought",
          summary: [{ type: "text", text: part.reasoning }],
          signature: OptionalStringSchema.parse(
            part.metadata?.thoughtSignature
          )
        });
      } else {
        const content = toInteractionContent(part);
        if (content) {
          normalContent.push(content);
        }
      }
    }
    if (normalContent.length > 0) {
      if (message.role === "model") {
        steps.push({
          type: "model_output",
          content: normalContent
        });
      } else {
        steps.push({
          type: "user_input",
          content: normalContent
        });
      }
    }
  }
  return steps;
}
function fromInteractionContent(content) {
  switch (content.type) {
    case "text":
      return fromTextContent(content);
    case "image":
      return fromImageContent(content);
    case "audio":
    case "document":
      return fromMediaContent(content);
    case "video":
      return fromVideoContent(content);
    case "thought":
      return fromThoughtContent(content);
    case "function_call":
      return fromFunctionCallContent(content);
    case "function_result":
      return fromFunctionResultContent(content);
    default:
      return {
        custom: {
          unknownContent: content
        }
      };
  }
}
function maybeAddGeminiThoughtSignature(step, part) {
  let updatedPart = part;
  if ("signature" in step && step.signature) {
    updatedPart.metadata = {
      ...part.metadata ?? {},
      thoughtSignature: step.signature
    };
  }
  return updatedPart;
}
function fromGoogleSearchCall(step) {
  const part = {
    custom: {
      googleSearchCall: {
        id: step.id,
        arguments: step.arguments
      }
    }
  };
  return maybeAddGeminiThoughtSignature(step, part);
}
function fromGoogleSearchResult(step) {
  const part = {
    custom: {
      googleSearchResult: {
        callId: step.call_id,
        result: step.result
      }
    }
  };
  return maybeAddGeminiThoughtSignature(step, part);
}
function fromCodeExecutionCall(step) {
  const part = {
    custom: {
      executableCode: {
        code: step.arguments.code,
        language: step.arguments.language || "PYTHON"
      }
    }
  };
  part.metadata = { callId: step.id };
  return maybeAddGeminiThoughtSignature(step, part);
}
function fromCodeExecutionResult(step) {
  const part = {
    custom: {
      codeExecutionResult: {
        output: typeof step.result === "string" ? step.result : JSON.stringify(step.result),
        outcome: "OUTCOME_OK"
      }
    }
  };
  part.metadata = { callId: step.call_id };
  return maybeAddGeminiThoughtSignature(step, part);
}
function fromServerFunctionCall(step) {
  return {
    custom: {
      serverFunctionCall: {
        id: step.id,
        name: step.name,
        arguments: step.arguments
      }
    }
  };
}
function fromServerFunctionResult(step) {
  return {
    custom: {
      serverFunctionResult: {
        callId: step.call_id,
        name: step.name,
        result: step.result,
        isError: step.is_error
      }
    }
  };
}
function fromInteractionStep(step) {
  switch (step.type) {
    case "model_output":
      return step.content.map(fromInteractionContent);
    case "user_input":
      return [];
    case "google_search_call":
      return [fromGoogleSearchCall(step)];
    case "google_search_result":
      return [fromGoogleSearchResult(step)];
    case "code_execution_call":
      return [fromCodeExecutionCall(step)];
    case "code_execution_result":
      return [fromCodeExecutionResult(step)];
    case "thought":
      return [fromThoughtContent(step)];
    case "function_call":
      return [fromServerFunctionCall(step)];
    case "function_result":
      return [fromServerFunctionResult(step)];
  }
  return [{ custom: { unknownStep: step } }];
}
function fromMediaContent(content) {
  let url = content.uri;
  if (content.data && content.mime_type) {
    url = `data:${content.mime_type};base64,${content.data}`;
  }
  return {
    media: {
      url: url || "",
      contentType: content.mime_type
    }
  };
}
function fromTextContent(content) {
  return {
    text: content.text || "",
    metadata: {
      annotations: content.annotations
    }
  };
}
function fromImageContent(content) {
  const part = fromMediaContent(content);
  if (content.resolution !== void 0) {
    part.metadata = { resolution: content.resolution };
  }
  return part;
}
function fromVideoContent(content) {
  const part = fromMediaContent(content);
  if (content.resolution !== void 0) {
    part.metadata = { resolution: content.resolution };
  }
  return part;
}
function fromThoughtContent(content) {
  let reasoning = "";
  if (content.summary) {
    reasoning = content.summary.map((c) => {
      if (c.type === "text") return c.text;
      return "[Image]";
    }).join("\n");
  }
  return {
    reasoning,
    metadata: {
      thoughtSignature: content.signature
    },
    custom: {
      thought: content
    }
  };
}
function fromFunctionCallContent(content) {
  return {
    toolRequest: {
      name: content.name,
      input: content.arguments,
      ref: content.id
    }
  };
}
function fromFunctionResultContent(content) {
  return {
    toolResponse: {
      name: content.name,
      output: content.result,
      ref: content.call_id
    }
  };
}
function fromInteractionSync(interaction) {
  if (interaction.status === "failed") {
    throw new Error("Interaction failed");
  }
  const response = {
    finishReason: "stop",
    message: {
      role: "model",
      content: [],
      ...interaction.id || interaction.environment_id ? {
        metadata: {
          ...interaction.id ? { interactionId: interaction.id } : {},
          ...interaction.environment_id ? { environmentId: interaction.environment_id } : {}
        }
      } : {}
    },
    custom: interaction,
    raw: interaction
  };
  if (interaction.status === "cancelled") {
    response.finishReason = "aborted";
    response.finishMessage = "Operation cancelled";
    response.message.content = [{ text: "Operation cancelled." }];
    return response;
  }
  const steps = interaction.steps;
  if (steps?.length) {
    response.message.content = steps.flatMap(fromInteractionStep).filter((p) => p && Object.keys(p).length > 0);
    if (interaction.usage) {
      response.usage = {
        inputTokens: interaction.usage.total_input_tokens,
        outputTokens: interaction.usage.total_output_tokens,
        totalTokens: interaction.usage.total_tokens,
        cachedContentTokens: interaction.usage.total_cached_tokens,
        thoughtsTokens: interaction.usage.total_thought_tokens
      };
      if (interaction.usage.input_tokens_by_modality) {
        for (const modalityToken of interaction.usage.input_tokens_by_modality) {
          switch (modalityToken.modality) {
            case "text":
              response.usage.inputCharacters = modalityToken.tokens;
              break;
            case "image":
              response.usage.inputImages = modalityToken.tokens;
              break;
            case "audio":
              response.usage.inputAudioFiles = modalityToken.tokens;
              break;
          }
        }
      }
      if (interaction.usage.output_tokens_by_modality) {
        for (const modalityToken of interaction.usage.output_tokens_by_modality) {
          switch (modalityToken.modality) {
            case "text":
              response.usage.outputCharacters = modalityToken.tokens;
              break;
            case "image":
              response.usage.outputImages = modalityToken.tokens;
              break;
            case "audio":
              response.usage.outputAudioFiles = modalityToken.tokens;
              break;
          }
        }
      }
    }
  }
  return response;
}
function fromInteraction(interaction, clientOptions) {
  const op = { id: interaction.id };
  if (clientOptions) {
    op.metadata = { clientOptions };
  }
  if (interaction.status === "in_progress") {
    op.done = false;
  } else if (interaction.status === "cancelled") {
    op.done = true;
    op.output = {
      finishReason: "aborted",
      finishMessage: "Operation cancelled",
      message: {
        role: "model",
        content: [{ text: "Operation cancelled." }],
        ...interaction.id || interaction.environment_id ? {
          metadata: {
            ...interaction.id ? { interactionId: interaction.id } : {},
            ...interaction.environment_id ? { environmentId: interaction.environment_id } : {}
          }
        } : {}
      }
    };
  } else if (interaction.status === "completed") {
    op.done = true;
    const steps = interaction.steps;
    if (steps?.length) {
      const content = steps.flatMap(fromInteractionStep).filter((p) => p && Object.keys(p).length > 0);
      op.output = {
        finishReason: "stop",
        message: {
          role: "model",
          content,
          ...interaction.id || interaction.environment_id ? {
            metadata: {
              ...interaction.id ? { interactionId: interaction.id } : {},
              ...interaction.environment_id ? { environmentId: interaction.environment_id } : {}
            }
          } : {}
        },
        custom: interaction,
        raw: interaction
      };
      if (interaction.usage) {
        op.output.usage = {
          inputTokens: interaction.usage.total_input_tokens,
          outputTokens: interaction.usage.total_output_tokens,
          totalTokens: interaction.usage.total_tokens,
          cachedContentTokens: interaction.usage.total_cached_tokens,
          thoughtsTokens: interaction.usage.total_thought_tokens
        };
        if (interaction.usage.input_tokens_by_modality) {
          for (const modalityToken of interaction.usage.input_tokens_by_modality) {
            switch (modalityToken.modality) {
              case "text":
                op.output.usage.inputCharacters = modalityToken.tokens;
                break;
              case "image":
                op.output.usage.inputImages = modalityToken.tokens;
                break;
              case "audio":
                op.output.usage.inputAudioFiles = modalityToken.tokens;
                break;
            }
          }
        }
        if (interaction.usage.output_tokens_by_modality) {
          for (const modalityToken of interaction.usage.output_tokens_by_modality) {
            switch (modalityToken.modality) {
              case "text":
                op.output.usage.outputCharacters = modalityToken.tokens;
                break;
              case "image":
                op.output.usage.outputImages = modalityToken.tokens;
                break;
              case "audio":
                op.output.usage.outputAudioFiles = modalityToken.tokens;
                break;
            }
          }
        }
      }
    }
  }
  return op;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ensureToolIds,
  fromCodeExecutionCall,
  fromCodeExecutionResult,
  fromGoogleSearchCall,
  fromGoogleSearchResult,
  fromInteraction,
  fromInteractionContent,
  fromInteractionStep,
  fromInteractionSync,
  fromServerFunctionCall,
  fromServerFunctionResult,
  toInteractionContent,
  toInteractionRole,
  toInteractionSteps,
  toInteractionTool
});
