// @genkit-ai/google-genai/lib/googleai/lyria.mjs
import { modelActionMetadata, z } from "genkit";
import { modelRef } from "genkit/model";
import { model as pluginModel } from "genkit/plugin";
import { isKnownKey } from "../common/utils.mjs";
import { createInteraction } from "./client.mjs";
import {
  ensureToolIds,
  fromInteractionSync,
  toInteractionSteps
} from "./interaction-converters.mjs";
import {
  calculateApiKey,
  checkApiKey,
  checkModelName,
  extractVersion,
  modelName
} from "./utils.mjs";
const LyriaConfigSchema = z.object({
  apiKey: z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  responseModalities: z.array(z.enum(["TEXT", "IMAGE", "AUDIO"])).describe(
    "The modalities to be used in response. Defaults to AUDIO and TEXT for Lyria."
  ).optional()
}).passthrough();
function commonRef(name, info, configSchema = LyriaConfigSchema) {
  return modelRef({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: false,
        media: true,
        // Lyria supports images as multimodal input
        tools: false,
        toolChoice: false,
        systemRole: false,
        output: ["media", "text"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("lyria-3");
const KNOWN_MODELS = {
  "lyria-3-clip-preview": commonRef("lyria-3-clip-preview"),
  "lyria-3-pro-preview": commonRef("lyria-3-pro-preview")
};
function isLyriaModelName(value) {
  return !!value?.startsWith("lyria-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (isKnownKey(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return modelRef({
    name: `googleai/${name}`,
    config,
    configSchema: LyriaConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isLyriaModelName(modelName(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return modelActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return pluginModel(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request) => {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        responseModalities,
        ...rest
      } = request.config || {};
      const apiKey = calculateApiKey(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion
      };
      const messages = structuredClone(request.messages);
      let responseModalitiesConverted = ["audio", "text"];
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        model: extractVersion(ref),
        input: toInteractionSteps(ensureToolIds(messages)),
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await createInteraction(apiKey, req, newClientOptions);
      return fromInteractionSync(response);
    }
  );
}
const TEST_ONLY = { KNOWN_MODELS };
export {
  LyriaConfigSchema,
  TEST_ONLY,
  defineModel,
  isLyriaModelName,
  listActions,
  listKnownModels,
  model
};
