// @genkit-ai/google-genai/lib/googleai/lyria.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var lyria_exports = {};
__export(lyria_exports, {
  LyriaConfigSchema: () => LyriaConfigSchema,
  TEST_ONLY: () => TEST_ONLY,
  defineModel: () => defineModel,
  isLyriaModelName: () => isLyriaModelName,
  listActions: () => listActions,
  listKnownModels: () => listKnownModels,
  model: () => model
});
module.exports = __toCommonJS(lyria_exports);
var import_genkit = require("genkit");
var import_model = require("genkit/model");
var import_plugin = require("genkit/plugin");
var import_utils = require("../common/utils.js");
var import_client = require("./client.js");
var import_interaction_converters = require("./interaction-converters.js");
var import_utils2 = require("./utils.js");
const LyriaConfigSchema = import_genkit.z.object({
  apiKey: import_genkit.z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: import_genkit.z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  responseModalities: import_genkit.z.array(import_genkit.z.enum(["TEXT", "IMAGE", "AUDIO"])).describe(
    "The modalities to be used in response. Defaults to AUDIO and TEXT for Lyria."
  ).optional()
}).passthrough();
function commonRef(name, info, configSchema = LyriaConfigSchema) {
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: false,
        media: true,
        // Lyria supports images as multimodal input
        tools: false,
        toolChoice: false,
        systemRole: false,
        output: ["media", "text"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("lyria-3");
const KNOWN_MODELS = {
  "lyria-3-clip-preview": commonRef("lyria-3-clip-preview"),
  "lyria-3-pro-preview": commonRef("lyria-3-pro-preview")
};
function isLyriaModelName(value) {
  return !!value?.startsWith("lyria-");
}
function model(version, config = {}) {
  const name = (0, import_utils2.checkModelName)(version);
  if ((0, import_utils.isKnownKey)(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return (0, import_model.modelRef)({
    name: `googleai/${name}`,
    config,
    configSchema: LyriaConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isLyriaModelName((0, import_utils2.modelName)(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return (0, import_genkit.modelActionMetadata)({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  (0, import_utils2.checkApiKey)(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
    experimental_debugTraces: pluginOptions?.experimental_debugTraces
  };
  return (0, import_plugin.model)(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request) => {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        responseModalities,
        ...rest
      } = request.config || {};
      const apiKey = (0, import_utils2.calculateApiKey)(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion
      };
      const messages = structuredClone(request.messages);
      let responseModalitiesConverted = ["audio", "text"];
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        model: (0, import_utils2.extractVersion)(ref),
        input: (0, import_interaction_converters.toInteractionSteps)((0, import_interaction_converters.ensureToolIds)(messages)),
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await (0, import_client.createInteraction)(apiKey, req, newClientOptions);
      return (0, import_interaction_converters.fromInteractionSync)(response);
    }
  );
}
const TEST_ONLY = { KNOWN_MODELS };
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LyriaConfigSchema,
  TEST_ONLY,
  defineModel,
  isLyriaModelName,
  listActions,
  listKnownModels,
  model
});
